File: app/aktors/messages/Testrun.java
package aktors.messages;

import akka.actor.ActorRef;
import org.bson.types.ObjectId;
import play.api.libs.json.JsObject;
import play.api.libs.json.JsString$;
import play.api.libs.json.JsValue;
import scala.Tuple2;
import scala.collection.JavaConversions;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Patrick Robinson on 20.04.15.
 */
public class Testrun {
    public ObjectId id;
    public Testplan testplan;
    public List<ActorRef> subscribers;

    public static Testrun fromJSON(JsObject run) throws MalformedURLException {
        Testrun result = new Testrun();
        result.id = new ObjectId(run.$bslash("id").toString());
	    if(run.$bslash("testplan") instanceof JsObject) result.testplan = Testplan.fromJSON((JsObject)run.$bslash("testplan"));
        return result;
    }

	public JsObject toJSON() {
		return toJSON(true);
	}

	public JsObject toJSON(boolean withPlan) {
		ArrayList<Tuple2<String, JsValue>> tuplesj = new ArrayList<>();
		tuplesj.add(Tuple2.apply("id", JsString$.MODULE$.apply(id.toString())));
		if(withPlan) tuplesj.add(Tuple2.apply("testplan", testplan.toJSON()));
		return new JsObject(JavaConversions.asScalaBuffer(tuplesj));
	}

	@Override
	public int hashCode() {
		return id.hashCode();
	}

	@Override
	public boolean equals(Object obj) {
        return
            ((obj instanceof Testrun) && ((Testrun)obj).id.equals(id))
        ||  ((obj instanceof JsObject) && new ObjectId(((JsObject)obj).$bslash("id").toString()).equals(id));
	}
}

File: app/aktors/messages/Testrun.scala
package aktors.messages

import java.util
import java.util.concurrent.TimeUnit

import akka.actor.ActorRef
import helper.JSONHelper
import org.bson.types.ObjectId
import play.api.libs.json._
import java.net.MalformedURLException
import java.util.List

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, Future}
import scala.concurrent.ExecutionContext.Implicits.global

/**
 * Created by Patrick Robinson on 20.04.15.
 */
object Testrun {
	@throws(classOf[MalformedURLException])
	def fromJSON(run: JsObject): Testrun = new Testrun(
		ID = new ObjectId(JSONHelper.JsStringToString(run.\("id")))
	,   Testplan = Future {if(run.\("testplan").isInstanceOf[JsObject]) Testplan.fromJSON(run.\("testplan").asInstanceOf[JsObject]) else null}
	)
}

class Testrun(ID: ObjectId = new ObjectId, Subscribers: List[ActorRef] = new util.ArrayList[ActorRef](), Testplan: Future[Testplan] = null) {
	def this(ID:ObjectId, Subscribers: List[ActorRef], Testplan:Testplan) = this(ID, Subscribers, Future {Testplan})

	var _testplan: Future[Testplan] = Testplan

	def getID: ObjectId = ID
	def getTestplan: Testplan = Await.result(Testplan, Duration(10, TimeUnit.MINUTES))
	def setTestplan(plan: Testplan) = _testplan = Future {plan}
	def setTestplan(plan: Future[Testplan]) = _testplan = plan
	def getSubscribers: List[ActorRef] = Subscribers

	def toJSON: JsObject = toJSON(true)
	def toJSON(withPlan: Boolean): JsObject =
		if(withPlan)
			Json.obj(
				"id" -> JsString(ID.toString)
			,   "testplan" -> getTestplan.toJSON
			)
		else
			Json.obj("id" -> JsString(ID.toString))

	override def hashCode: Int = ID.hashCode

	override def equals(other: Any): Boolean = other match {
		case that: Testrun => getID.equals(that.getID)
		case json: JsObject => new ObjectId(JSONHelper.JsStringToString((json.\("id")))).equals(this.ID)
		case _ => false
	}
}